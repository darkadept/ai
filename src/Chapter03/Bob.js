import Genome from './Genome';
import {randomFloat, random, bits2int} from '../lib';

export default class Bob {
	constructor({
		crossoverRate,
		mutationRate,
		populationSize,
		chromoLength,
		geneLength,
	            },
	            map) {

		this._options = {
			populationSize,
			crossoverRate,
			mutationRate,
			chromoLength,
			geneLength,
		};

		this.genomes = [];
		this.generation = 0;
		this.fittestGenomeIndex = 0;
		this.bestFitnessScore = 0;
		this.totalFitnessScore = 0;
		this._running = true;

		this.map = map;

		this.createStartPopulation();
	}

	get running() {return this._running;}

	/**
	 * Select member of population by using roulette wheel selection
	 * @return {*}
	 */
	rouletteWheelSelection() {
		const {populationSize} = this._options;
		const slice = randomFloat() * this.totalFitnessScore;
		let total = 0.0;
		let selectedGenome = 0; // TODO picks first genome if none better, maybe should be random
		for (let i=0; i<populationSize; i++) {
			total += this.genomes[i].fitness;
			if (total > slice) {
				selectedGenome = i;
				break;
			}
		}
		return this.genomes[selectedGenome];
	}

	mutate(vecBits) {
		for (let curBit=0; curBit<vecBits.length; curBit++) {
			if (randomFloat() < this._mutationRate) {
				// modify vecBits
			}
		}
	}

	crossover(mum, dad) {
		if (randomFloat() > this._crossoverRate || mum.equals(dad)) {
			return {
				baby1: mum,
				baby2: dad,
			}
		}

		const cp = random(0, this._chromoLength - 1);

		// for (let i=0; i<cp; i++) {
		// 	baby1.
		// }


	}

	createStartPopulation() {
		const {populationSize, chromoLength} = this._options;

		// Clear the current population
		this.genomes = [];

		// Add a new population of random genomes
		for (let i = 0; i < populationSize; i++) {
			this.genomes.push(new Genome(chromoLength));
		}

		// Reset variables
		this.generation = 0;
		this.fittestGenome = 0;
		this.bestFitnessScore = 0;
		this.totalFitnessScore = 0;
	}

	epoch() {
		console.log('Epoch');
		const {populationSize} = this._options;

		this.updateFitnessScores();

		let newBabies = 0;
		let babyGenomes = [];
		while (newBabies < populationSize) {
			const mum = this.rouletteWheelSelection();
			const dad = this.rouletteWheelSelection();

		// 	let {baby1, baby2} = this.crossover(mum, dad);

		// 	baby1 = this.mutate(baby1);
		// 	baby2 = this.mutate(baby2);

		// 	babyGenomes.push(baby1);
		// 	babyGenomes.push(baby2);

			newBabies += 2;
		}
		//
		// this._genomes = babyGenomes;
		// this._generation += 1;
	}

	updateFitnessScores() {
		const {populationSize} = this._options;

		this.fittestGenome = 0;
		this.bestFitnessScore = 0;
		this.totalFitnessScore = 0;

		for (let i=0; i<populationSize; i++) {
			// Generate array of direction codes from the genome bits
			const directions = this.decode(this.genomes[i]);

			// Test the directions against the map, storing the fitness value
			this.genomes[i].fitness = this.map.testRoute(directions);

			// Add total fitness
			this.totalFitnessScore += this.genomes[i].fitness;

			// If this is the fittest genome, store result
			if (this.genomes[i].fitness > this.bestFitnessScore) {
				this.bestFitnessScore = this.genomes[i].fitness;
				this.fittestGenome = i;

				// store temp memory from testing route

				// Check for exit
				if (this.genomes[i].fitness === 1) {
					this._running = false;
				}
			}
		}

		// reset temp memory
	}

	/**
	 * Decode genome bits to array of direction codes
 	 * @param genome - The genome to decode
	 * @return {Array} - An array of directions 0,1,2,3
	 */
	decode(genome) {
		const {geneLength} = this._options;
		let directions = [];

		// Jump through genome bits by gene length
		for (let gene=0; gene<genome.length; gene += geneLength) {
			// Get slice of the genome and convert bits to integer code
			const thisGene = genome.slice(gene, geneLength);
			directions.push(bits2int(thisGene));
		}
		return directions;
	}
}
