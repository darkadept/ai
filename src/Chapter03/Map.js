import {random, create2dArray} from '../lib';
import sha256 from '../sha256';

export default class Map {
	constructor(width, height) {
		this._map = create2dArray(width, height);
		this._width = width;
		this._height = height;

		this._start = {x: 0, y: 0};
		this._end = {x: 0, y:0};
	}

	/**
	 * Tests an array of direction integers and returns a fitness score.
	 * @param vecPath
	 * @return {number}
	 */
	testRoute(vecPath) {
		let pos = {x: this._start.x, y: this._start.y};

		vecPath.forEach(vec => {
			switch (vec) {
				case 0: // north
					if (pos.y-1 < 0 || this.map[pos.x][pos.y-1] === 1) break;
					pos.y = pos.y - 1;
					break;
				case 1: // south
					if (pos.y+1 >= this._height || this.map[pos.x][pos.y+1] === 1) break;
					pos.y = pos.y + 1;
					break;
				case 2: // east
					if (pos.x+1 >= this._width || this.map[pos.x+1][pos.y] === 1) break;
					pos.x = pos.x + 1;
					break;
				case 3: // west
					if (pos.x-1 < 0 || this._map[pos.x-1][pos.y] === 1) break;
					pos.x = pos.x - 1;
					break;
			}

			// TODO Check if we already found the exit here.
		});

		// Calculate fitness score (0 to 1)
		const diff = {
			x: Math.abs(pos.x - this._end.x),
			y: Math.abs(pos.y - this._end.y),
		};
		return 1 / (diff.x + diff.y + 1);
	}

	get width() {return this._width};
	get height() {return this._height};
	get map() {
		return this._map;
	}
	get start() {return this._start};
	get end() {return this._end};

	/**
	 * Returns a map filled with Konva objects
	 */
	get renderMap() {
		let map = create2dArray(this._width, this._height);
		for (let x=0; x<this._width; x++) {
			for (let y=0; y<this._height; y++) {
				if (this._map[x][y]) map[x][y] = {fill: 'red'};
			}
		}

		map[this._start.x][this._start.y] = {fill: 'green'};
		map[this._end.x][this._end.y] = {fill: 'blue'};

		return map;
	}

	get hash() {
		return sha256(JSON.stringify(this._map));
	}

	/**
	 * Generates new random map data
	 * @param percent
	 * @return {Map}
	 */
	generateMap(percent=80) {
		for (let x=0; x<this._width; x++) {
			for (let y=0; y<this._height; y++) {
				if (random() > percent) {
					this._map[x][y] = 1;
				}
			}
		}

		this._start.x = random(0, this._width-1);
		this._start.y = random(0, this._height-1);
		this._map[this._start.x][this._start.y] = 0; // Make sure the start is accessible

		this._end.x = random(0, this._width-1);
		this._end.y = random(0, this._height-1);
		this._map[this._end.x][this._end.y] = 0; // Make sure the end is accessible

		return this;
	}

	loadMap(map, start, end) {
		this._map = map;
		this._start = start;
		this._end = end;

		return this;
	}
}
