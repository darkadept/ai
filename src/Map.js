import {random, create2dArray} from './lib';

export default class Map {
	constructor(width, height) {
		this._map = create2dArray(width, height);
		this._width = width;
		this._height = height;

		this._start = {x: 0, y: 0};
		this._end = {x: 0, y:0};
	}

	testRoute(vecPath) {
		let pos = {x: this._start.x, y: this._start.y};

		vecPath.forEach(vec => {
			switch (vec) {
				case 0: // north
					if (pos.y-1 < 0 || this.map[pos.x][pos.y-1] === 1) break;
					pos.y = pos.y - 1;
					break;
				case 1: // south
					if (pos.y+1 >= this._height || this.map[pos.x][pos.y+1] === 1) break;
					pos.y = pos.y + 1;
					break;
				case 2: // east
					if (pos.x+1 >= this._width || this.map[pos.x+1][pos.y] === 1) break;
					pos.x = pos.x + 1;
					break;
				case 3: // west
					if (pos.x-1 < 0 || this._map[pos.x-1][pos.y] === 1) break;
					pos.x = pos.x - 1;
					break;
			}
		});

		const diff = {
			x: Math.abs(pos.x - this._end.x),
			y: Math.abs(pos.y - this._end.y),
		};

		return 1 / (diff.x + diff.y + 1);
	}

	get width() {return this._width};
	get height() {return this._height};

	get map() {
		return this._map;
	}

	get renderMap() {
		let map = create2dArray(this._width, this._height);
		for (let x=0; x<this._width; x++) {
			for (let y=0; y<this._height; y++) {
				if (this._map[x][y]) map[x][y] = {fill: 'red'};
			}
		}

		map[this._start.x][this._start.y] = {fill: 'green'};
		map[this._end.x][this._end.y] = {fill: 'blue'};

		return map;
	}

	generateMap(percent=80) {
		for (let x=0; x<this._width; x++) {
			for (let y=0; y<this._height; y++) {
				if (random() > percent) {
					this._map[x][y] = 1;
				}
			}
		}

		this._start.x = random(0, this._width-1);
		this._start.y = random(0, this._height-1);

		this._end.x = random(0, this._width-1);
		this._end.y = random(0, this._height-1);
	}
}
