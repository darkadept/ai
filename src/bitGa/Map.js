import {random, create2dArray} from '../lib';

export default class Map {
	constructor({
    width,
    height,
		map = null,
		start = {x: 0, y: 0},
		end = {x: 0, y: 0},
	}) {
		this._map = map || create2dArray(width, height);
		this._width = width;
		this._height = height;

		this._start = start;
		this._end = end;

		this._id = random(0, 999999);
	}

	get start() {return this._start;}
	get end() {return this._end;}
	get id() {return this._id;}

	at(x, y) {
		// Check bounds
		if (x < 0 || y < 0 || x >= this._width || y >= this._height) return 1;
		// Return what is at the postition
		return this._map[x][y];
	}

	/**
	 * Randomly generate a map and start and end points.
	 * @param percentFill
	 * @returns {Map}
	 */
	generate(percentFill=80) {
		for (let x=0; x<this._width; x++) {
			for (let y=0; y<this._height; y++) {
				if (random() > percentFill) {
					this._map[x][y] = 1;
				}
			}
		}

		this._start.x = random(0, this._width-1);
		this._start.y = random(0, this._height-1);
		this._map[this._start.x][this._start.y] = 0; // Make sure the start is accessible

		this._end.x = random(0, this._width-1);
		this._end.y = random(0, this._height-1);
		this._map[this._end.x][this._end.y] = 0; // Make sure the end is accessible

		return this;
	}

	/**
	 * Returns the map data as a single structure.
	 * @returns {{map: *, width: *, height: *, start: *, end: *}}
	 */
	save() {
		return {
			map: this._map,
			width: this._width,
			height: this._height,
			start: this._start,
			end: this._end,
		};
	}
}
