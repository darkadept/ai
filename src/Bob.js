import Genome from './Genome';
import Map from './Map';

export default class Bob {
	constructor() {
		this._genomes = [];
		this._population = 0;
		this._crossoverRate = 0;
		this._mutationRate = 0;
		this._chromoLength = 0;
		this._fittestGenome = null;
		this._bestFitnessScore = 0;
		this._totalFitnessSCore = 0;
		this._generation = 0;

		// this._map = new Map()
	}
}
