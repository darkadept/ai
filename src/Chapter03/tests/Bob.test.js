import Bob from '../Bob';
import Map from '../Map';

const mockMath = Object.create(global.Math);
let randomFlip = true;
mockMath.random = () => {
	if (randomFlip) {
		randomFlip = false;
		return 0.3;
	}
	randomFlip = true;
	return 0.6;
};
global.Math = mockMath;

const map = new Map(5, 3);

const options = {
	crossoverRate: 0.7,
	mutationRate: 0.001,
	populationSize: 10,
	chromoLength: 70,
	geneLength: 2,
};

it('should construct a new population on initialize', () => {
	const bob = new Bob(options, map);
	expect(bob.genomes).toMatchSnapshot();
});

it('should update fitness scores', () => {
	const bob = new Bob(options, map);
	bob.updateFitnessScores();
	expect(bob.genomes).toMatchSnapshot();
	expect(bob.totalFitnessScore).toBeCloseTo(0.33333, 2);
});
