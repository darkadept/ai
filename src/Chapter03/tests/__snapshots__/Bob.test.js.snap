// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should construct a new population on initialize 1`] = `
Array [
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
]
`;

exports[`should update fitness scores 1`] = `
Array [
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
  Genome {
    "_fitness": 0.3333333333333333,
    "_vecBits": Array [
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
      0,
      1,
    ],
  },
]
`;
